coral-incremental/src/main/java/com/linkedin/coral/incremental/RelNodeCostEstimator.java

```diff
@@ -0,0 +1,237 @@
+    this.IOCostParam = IOCostParam;
+  }
+
+  public void setShuffleCostParam(Double shuffleCostParam) {
+    this.shuffleCostParam = shuffleCostParam;
+  }
+
+  public void loadStatistic(String configPath) {
+    try {
+      String content = new String(Files.readAllBytes(Paths.get(configPath)));
+      // Parse JSON string to JsonObject
+      JsonObject jsonObject = new JsonParser().parse(content).getAsJsonObject();
+      // Iterate over each table in the JSON object
+      for (Map.Entry<String, JsonElement> entry : jsonObject.entrySet()) {
+        String tableName = entry.getKey();
+        JsonObject tableObject = entry.getValue().getAsJsonObject();
+
+        // Extract row count
+        Double rowCount = tableObject.get("RowCount").getAsDouble();
+
+        // Extract distinct counts
+        JsonObject distinctCounts = tableObject.getAsJsonObject("DistinctCounts");
+
+        stat.put(tableName, rowCount);
+
+        // Iterate over distinct counts
+        for (Map.Entry<String, JsonElement> distinctEntry : distinctCounts.entrySet()) {
+          String columnName = distinctEntry.getKey();
+          Double distinctCount = distinctEntry.getValue().getAsDouble();
+
+          distinctStat.put(tableName + ":" + columnName, distinctCount);
+        }
+
+      }
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+
+  }
+
+  public Double getCost(RelNode rel) {
+    CostInfo executionCostInfo = getExecutionCost(rel);
+    Double IOCost = executionCostInfo.row * IOCostParam;
+    return executionCostInfo.cost * shuffleCostParam + IOCost;
+  }
+
+  public CostInfo getExecutionCost(RelNode rel) {
+    if (rel instanceof TableScan) {
+      return getExecutionCostTableScan((TableScan) rel);
+    } else if (rel instanceof LogicalJoin) {
+      return getExecutionCostJoin((LogicalJoin) rel);
+    } else if (rel instanceof LogicalUnion) {
+      return getExecutionCostUnion((LogicalUnion) rel);
+    } else if (rel instanceof LogicalProject) {
+      return getExecutionCostProject((LogicalProject) rel);
+    }
+    return new CostInfo(0.0, 0.0);
+  }
+
+  private CostInfo getExecutionCostTableScan(TableScan scan) {
+    RelOptTable originalTable = scan.getTable();
+    String tableName = getTableName(originalTable);
+    Double row = stat.getOrDefault(tableName, 5.0);
+    return new CostInfo(row, row);
+  }
+
+  private String getTableName(RelOptTable table) {
+    return String.join(".", table.getQualifiedName());
+  }
+
+  private CostInfo getExecutionCostJoin(LogicalJoin join) {
+    RelNode left = join.getLeft();
+    RelNode right = join.getRight();
+    if (!(left instanceof TableScan) || !(right instanceof TableScan)) {
+      return new CostInfo(0.0, 0.0);
+    }
+    CostInfo leftCost = getExecutionCost(left);
+    CostInfo rightCost = getExecutionCost(right);
+    Double joinSize = estimateJoinSize(join, leftCost.row, rightCost.row);
+    return new CostInfo(max(leftCost.cost, rightCost.cost), joinSize);
```

Comment:
Can you explain here and in comments why we are taking the max?

Reply:
Because in modern distributed systems like Spark, the shuffle cost is dominated by the largest shuffle, multicore execute the left and right at the same time.